Mathlib/Algebra/Algebra/Basic.lean

```diff
@@ -268,59 +274,109 @@ variable {R}
 instance int_algebra_subsingleton : Subsingleton (Algebra ℤ R) :=
   ⟨fun P Q => Algebra.algebra_ext P Q <| RingHom.congr_fun <| Subsingleton.elim _ _⟩
 
+@[simp]
+lemma algebraMap_comp_int_cast_eq_cast (R A : Type*) [CommRing R] [Ring A] [Algebra R A] :
+    algebraMap R A ∘ Int.cast = Int.cast := by
+  ext; simp
+
 end Int
 
-namespace NoZeroSMulDivisors
+section FaithfulSMul
 
-variable {R A : Type*}
+instance (R : Type*) [NonAssocRing R] : FaithfulSMul R R := ⟨fun {r₁ r₂} h ↦ by simpa using h 1⟩
 
-open Algebra
+variable (R A : Type*) [CommSemiring R] [Semiring A]
 
-/-- If `algebraMap R A` is injective and `A` has no zero divisors,
-`R`-multiples in `A` are zero only if one of the factors is zero.
+lemma faithfulSMul_iff_injective_smul_one [Module R A] [IsScalarTower R A A] :
+    FaithfulSMul R A ↔ Injective (fun r : R ↦ r • (1 : A)) := by
+  refine ⟨fun ⟨h⟩ {r₁ r₂} hr ↦ h fun a ↦ ?_, fun h ↦ ⟨fun {r₁ r₂} hr ↦ h ?_⟩⟩
+  · simp only at hr
+    rw [← one_mul a, ← smul_mul_assoc, ← smul_mul_assoc, hr]
+  · simpa using hr 1
 
-Cannot be an instance because there is no `Injective (algebraMap R A)` typeclass.
--/
-theorem of_algebraMap_injective [CommSemiring R] [Semiring A] [Algebra R A] [NoZeroDivisors A]
-    (h : Function.Injective (algebraMap R A)) : NoZeroSMulDivisors R A :=
-  ⟨fun hcx => (mul_eq_zero.mp ((smul_def _ _).symm.trans hcx)).imp_left
-    (map_eq_zero_iff (algebraMap R A) h).mp⟩
+variable [Algebra R A]
+
+lemma faithfulSMul_iff_algebraMap_injective :
+    FaithfulSMul R A ↔ Injective (algebraMap R A) := by
```

Comment:
There are a lot of other places where the statement could be one line ...
```suggestion
lemma faithfulSMul_iff_algebraMap_injective : FaithfulSMul R A ↔ Injective (algebraMap R A) := by
```

Reply:
Done (and elsewhere).